crates/network-scheduler/src/assignment.rs

```diff
@@ -0,0 +1,235 @@
+            Some(worker_assignment) => worker_assignment,
+            None => {
+                return None
+            }
+        };
+        let mut result: Vec<Dataset> = Default::default();
+        let mut idxs: Vec<u64> = Default::default();
+        let mut cursor = 0;
+        for v in &local_assignment.chunks_deltas {
+            cursor += v;
+            idxs.push(cursor);
+        }
+        cursor = 0;
+        for u in &self.datasets {
+            if idxs.is_empty() {
+                break;
+            }
+            let mut filtered_chunks: Vec<Chunk> = Default::default();
+            for v in &u.chunks {
+                if idxs[0] < cursor {
+                    return None; // Malformed diffs
+                }
+                if idxs[0] == cursor {
+                    filtered_chunks.push(v.clone());
+                    idxs.remove(0);
+                }
+                if idxs.is_empty() {
+                    break;
+                }
+                cursor += 1;
+            }
+            if !filtered_chunks.is_empty() {
+                result.push(Dataset {
+                    id: u.id.clone(),
+                    base_url: u.base_url.clone(),
+                    chunks: filtered_chunks
+                });
+            }
+        }
+        Some(result)
+        // let mut result: Vec<Chunk> = Default::default();
+        // let mut cursor = 0;
+        // for v in &local_assignment.chunks_deltas {
+        //     cursor += v;
+        //     result.push(flat_chunks[cursor as usize].clone());
+        // }
+        // Some(result)
+    }
+
+    pub fn headers_for_peer_id(&self, peer_id: String) -> Option<HashMap<String, String>> {
+        let local_assignment = match self.worker_assignments.get(&peer_id) {
+            Some(worker_assignment) => worker_assignment,
+            None => {
+                return None
+            }
+        };
+        let headers = match serde_json::to_value(&local_assignment.encrypted_headers) {
+            Ok(v) => v,
+            Err(_) => {
+                return None;
+            }
+        };
+        let mut result: HashMap<String, String> = Default::default();
+        for (k,v) in headers.as_object().unwrap() {
+            result.insert(k.to_string(), v.as_str().unwrap().to_string());
+        }
+        Some(result)
+    }
+
+    pub fn chunk_index(&mut self, chunk_id: String) -> Option<u64> {
+        if self.chunk_map == None {
+            let mut chunk_map: HashMap<String, u64> = Default::default();
+            let mut idx = 0;
+            for dataset in &self.datasets {
+                for chunk in &dataset.chunks {
+                    chunk_map.insert(chunk.id.clone(), idx);
+                    idx += 1;
+                }
+            };
+            self.chunk_map = Some(chunk_map);
+        };
+        match self.chunk_map.as_ref().unwrap().get(&chunk_id) {
+            Some(idx) =>Some(idx.clone()),
+            None => None
+        }
```

Comment:
There are quite a lot of useful clippy warnings

Reply:
fixed